### PR TITLE
Document class and module attributes

### LINKED ISSUE #31: Add class/instance and module attributes/properties documentation
### What's needed?

We need to properly document class/instance and module attributes/properties documentation so they appear in the generated docs.

### Proposed solution

For now how to best do this is documented in this issue:

* https://github.com/frequenz-floss/frequenz-channels-python/pull/25

### Use cases

_No response_

### Alternatives and workarounds

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by daniel-zullo-frequenz**
Add or amend docstrings for all classes and modules where needed.

Fixes #31

### CODE REVIEW

File: src/frequenz/channels/_anycast.py [hunk 1]
```diff
@@ -68,10 +68,19 @@ def __init__(self, maxsize: int = 10) -> None:
             maxsize: Size of the channel's buffer.
         """
         self.limit: int = maxsize
+        """The maximum number of items that can be stored in the channel's buffer."""
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/channels/_anycast.py]: ?
```suggestion
        """The maximum number of messages that can be stored in the channel's buffer."""
```

File: src/frequenz/channels/_anycast.py [hunk 2]
```diff
@@ -68,10 +68,19 @@ def __init__(self, maxsize: int = 10) -> None:
...
+
         self.deque: Deque[T] = deque(maxlen=maxsize)
+        """The channel's buffer."""
+
         self.send_cv: Condition = Condition()
+        """The condition to wait for free space in the channel's buffer."""
+
         self.recv_cv: Condition = Condition()
+        """The condition to wait for data in the channel's buffer."""
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/channels/_anycast.py]: ?
```suggestion
        """The condition to wait for messages in the channel's buffer."""
```

File: src/frequenz/channels/_base_classes.py [hunk 3]
```diff
@@ -169,11 +169,13 @@ def __init__(self, recv: Receiver[T], transform: Callable[[T], U]) -> None:
 
         Args:
             recv: The input receiver.
-            transform: The function to run on the input
-                data.
+            transform: The function to run on the input data.
         """
         self._recv = recv
+        """The input receiver."""
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/channels/_base_classes.py:175]:
```suggestion
        """The input channel receiver."""
```

File: src/frequenz/channels/_anycast.py (hunk 1 above)
Comment (llucax) [reply src/frequenz/channels/_anycast.py]: We generally call it a `value` (what's sent through the channels).

I would also mention explicitly what happens if the buffer is full. Looking at the code it seems the sender will block at `send()` until a message is consumed. Maybe @shsms likes to confirm, but if this is the case, this is relevant to #76, as we already have some senders that can block (this one) and some that never do (broadcast).

File: src/frequenz/channels/_anycast.py [hunk 4]
```diff
@@ -68,10 +68,19 @@ def __init__(self, maxsize: int = 10) -> None:
             maxsize: Size of the channel's buffer.
         """
         self.limit: int = maxsize
```
Comment (llucax) [inline src/frequenz/channels/_anycast.py:70]: Damn, all of these should be private, or read-only at best. Not for this PR though:

* #186

File: src/frequenz/channels/_base_classes.py (hunk 3 above)
Comment (llucax) [reply src/frequenz/channels/_base_classes.py:175]: I think just receiver is fine, as it could not be from a channel (like `Timer` or `FileWatcher`).

File: src/frequenz/channels/_bidirectional.py [hunk 5]
```diff
@@ -117,21 +122,29 @@ def __init__(self, client_id: str, service_id: str) -> None:
             service_id: A name for the service end of the channels.
         """
         self._client_id = client_id
+        """The name for the client, used to name the channels."""
```
Comment (llucax) [inline src/frequenz/channels/_bidirectional.py:125]: This is really confusing, but I know the original docs for `client_id` is and even the name of the variable.

But I guess we shouldn't fix the documentation, but fix the interface instead:

* #187

File: src/frequenz/channels/_broadcast.py [hunk 6]
```diff
@@ -85,12 +85,22 @@ def __init__(self, name: str, resend_latest: bool = False) -> None:
                 to wait for the next message on the channel to arrive.
         """
         self.name: str = name
+        """The name of the broadcast channel."""
```
Comment (llucax) [inline src/frequenz/channels/_broadcast.py]: I would mention that this is for debugging purposes.

File: src/frequenz/channels/_broadcast.py [hunk 7]
```diff
@@ -222,11 +233,19 @@ def __init__(self, uuid: UUID, name: str, maxsize: int, chan: Broadcast[T]) -> N
                 belongs to.
         """
         self._uuid = uuid
+        """The UUID to identify the receiver in the broadcast channel's list of receivers."""
+
         self._name = name
+        """The name to identify the receiver."""
```
Comment (llucax) [inline src/frequenz/channels/_broadcast.py]: I would also mention this is only for debugging.

File: src/frequenz/channels/_broadcast.py [hunk 8]
```diff
@@ -222,11 +233,19 @@ def __init__(self, uuid: UUID, name: str, maxsize: int, chan: Broadcast[T]) -> N
...
+
         self._chan = chan
+        """The broadcast channel that this receiver belongs to."""
+
         self._q: Deque[T] = deque(maxlen=maxsize)
+        """The receiver's internal message queue."""
 
         self._active = True
+        """Whether the receiver is still active."""
```
Comment (llucax) [inline src/frequenz/channels/_broadcast.py]: Maybe mention what happens when it is inactive, or how/when it becomes inactive? After a quick look it seems the only way to make it inactive is by converting it into a peekable.

**Review by llucax — COMMENTED** [review]
Just a few small comments and a few improvements that we might want to make before 1.0 as they are breaking changes :disappointed: (I created new issues for them).

File: src/frequenz/channels/_base_classes.py (hunk 3 above)
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/_base_classes.py:175]: Good point, then will keep it as it is and resolving this conversation...

File: src/frequenz/channels/_anycast.py (hunk 4 above)
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/_anycast.py:70]: I see. Thanks for creating https://github.com/frequenz-floss/frequenz-channels-python/issues/186

File: src/frequenz/channels/_anycast.py (hunk 1 above)
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/_anycast.py]: btw `self.limit` is no longer used/needed so it can actually be deleted. Only `maxsize` is used to initialize the channel's buffer. I'll remove it if there is no objections.
And I'd extend the documentation for the blocking conditions instead

File: src/frequenz/channels/_broadcast.py (hunk 8 above)
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/_broadcast.py]: done

File: src/frequenz/channels/_anycast.py (hunk 1 above)
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/_anycast.py]: Added a new commit to remove `self.limit`

File: src/frequenz/channels/_anycast.py [hunk 9]
```diff
@@ -67,19 +67,35 @@ def __init__(self, maxsize: int = 10) -> None:
         Args:
             maxsize: Size of the channel's buffer.
         """
-        self.limit: int = maxsize
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/channels/_anycast.py:70]: this needs an entry into the release notes

File: RELEASE_NOTES.md [hunk 10]
```diff
@@ -6,7 +6,7 @@ The `Timer` now can be started with a delay.
 
 ## Upgrading
 
-<!-- Here goes notes on how to upgrade from previous versions, including deprecations and what they should be replaced with -->
+* The `limit` attribute was removed from `Anycast` channel as the same information can be obtained from its buffer through `deque.maxlen` property.
```
Comment (llucax) [inline RELEASE_NOTES.md]: Why is this removed? If it is because of #186, then it should be made read-only, not removed, and I would leave it for a separate PR.

**Review by llucax — COMMENTED** [review]
Not sure why did you remove the `limit` in the PR. The rest LGTM.

Also, could you not mark conversations I started as resolved? If you do so, I need to go expanding them one by one to see your replies and to remember what comments I made to go see in the code how it was addressed. If you need to keep track of which comments you already addressed, you can use a reaction instead, like :+1: .

File: RELEASE_NOTES.md (hunk 10 above)
Comment (daniel-zullo-frequenz) [reply RELEASE_NOTES.md]: It wasn't because of #186. Because of the comment I've found it redundant as the variable was only set and actually never used in the channel itself. Then exactly the same information can be retrieve from the `deque.maxlen` property. I've added a separate commit to remove it so I can easily discard the commit if you're against this change

Comment (llucax) [reply RELEASE_NOTES.md]: I think in the future I wouldn't give access to users to the underlying `deque`, but it is nice to still provide a way to know what's the buffer size. In any case, I wouldn't decide this in this PR, we need to think about it a bit more before changing the API.

Comment (daniel-zullo-frequenz) [reply RELEASE_NOTES.md]: fair enough, then I'll get rid of the related commits were `limit` was removed

Comment (daniel-zullo-frequenz) [reply RELEASE_NOTES.md]: done

**Review by llucax — APPROVED** [review]
Thanks!
